src/coreclr/inc/corinfo.h

```diff
@@ -2403,6 +2441,42 @@ class ICorStaticInfo
             int32_t                     num
             ) = 0;
 
+    //------------------------------------------------------------------------------
+    // getTypeLayout: Obtain a tree describing the layout of a type.
+    //
+    // Parameters:
+    //   typeHnd            - Handle of the type.
+    //   treeNodes          - [out] Pointer to tree node entries to write.
```

Comment:
I see in the JIT you call this with an arbitrary array of 256 elements, and bail on numbers above that.

If a caller wanted to allocate exactly the right number of nodes, what would they do? Perhaps they could 
call this with `treeNodes == nullptr` and `*numTreeNodes == 0` and on exit `*numTreeNodes` would equal the size needed? But that is semantically questionable since `*numTreeNodes` is "the number of entries written" which would have to be zero. So you'd have to iterate, sending larger and larger buffers to the API.

Should it be supported to get exactly the number needed in one call? Either here, or in a `getTypeLayoutNodeCount` API?

Reply:
> Technically `[in, out]`? Specifically, on entry it points to an array of `CORINFO_TYPE_LAYOUT_NODE` memory with `*numTreeNodes` elements.

I'll add a comment to this regard.

> I see in the JIT you call this with an arbitrary array of 256 elements, and bail on numbers above that.
> 
> If a caller wanted to allocate exactly the right number of nodes, what would they do? Perhaps they could
call this with treeNodes == nullptr and *numTreeNodes == 0 and on exit *numTreeNodes would equal the size needed? But that is semantically questionable since *numTreeNodes is "the number of entries written" which would have to be zero. So you'd have to iterate, sending larger and larger buffers to the API.
> 
> Should it be supported to get exactly the number needed in one call? Either here, or in a getTypeLayoutNodeCount API?

I picked the number by measuring how many nodes the largest structs in all our SPMI collections have, doubling it, and rounding it up (IIRC it was between 80-100 nodes).

The 256 limit on the JIT side also serves to limit ourselves from pathological cases, because it is very easy (especially with upcoming inline arrays) to create structs with tons and tons of nodes in its layout. For physical promotion hitting this limit will just result in slightly conservative behavior, where we will believe padding to be significant.  For regular promotion we have a natural limit of 9 nodes due to its restrictions. So I would say that for now this is ok, and if we end up in a situation where it is too conservative in physical promotion we can look into improving it (IMO, a more preferable design in that situation might be callback based, e.g. `walkTypeLayout` -- but not sure we have much prior art to that).

Reply:
I suppose the API could take a range to return fields in, though then I think again the callback approach would be more flexible and with fewer downsides.

The actual "field type" parts are not used by physical promotion, so if regular promotion didn't exist this API would be quite a bit simpler (returning just a flat array of the significant segments). But for regular promotion it would technically be fine if it picked out a set of up to 4 fields it is interested in and only tried to promote these, though of course we would end up with dependent promotion a lot of the time.